File: Reinforcement-Learning/q_learning.py
# -*- coding: utf-8 -*-
import random
from abc import ABCMeta, abstractmethod
from operator import itemgetter


class QLearning(metaclass=ABCMeta):
    '''
    Abstract base class and `Template Method Pattern` of Q-Learning.

    Attributes:
        alpha_value:        Learning rate.
        gamma_value:        Gammma value.
        q_dict:             Q(state, action) 
        r_dict:             R(state)
        t:                  time.

    '''

    # Learning rate.
    __alpha_value = 0.1

    def get_alpha_value(self):
        '''
        getter
        Learning rate.
        '''
        if isinstance(self.__alpha_value, float) is False:
            raise TypeError("The type of __alpha_value must be float.")
        return self.__alpha_value

    def set_alpha_value(self, value):
        '''
        setter
        Learning rate.
        '''
        if isinstance(value, float) is False:
            raise TypeError("The type of __alpha_value must be float.")
        self.__alpha_value = value

    alpha_value = property(get_alpha_value, set_alpha_value)

    # Gamma value.
    __gamma_value = 0.5

    def get_gamma_value(self):
        '''
        getter
        Gamma value.
        '''
        if isinstance(self.__gamma_value, float) is False:
            raise TypeError("The type of __gamma_value must be float.")
        return self.__gamma_value

    def set_gamma_value(self, value):
        '''
        setter
        Gamma value.
        '''
        if isinstance(value, float) is False:
            raise TypeError("The type of __gamma_value must be float.")
        self.__gamma_value = value

    gamma_value = property(get_gamma_value, set_gamma_value)

    # Q(state, action)
    __q_dict = {}

    def get_q_dict(self):
        '''
        getter
        Q(state, action)
        '''
        if isinstance(self.__q_dict, dict) is False:
            raise TypeError("The type of __q_dict must be dict.")
        return self.__q_dict

    def set_q_dict(self, value):
        '''
        setter
        Q(state, action)
        '''
        if isinstance(value, dict) is False:
            raise TypeError("The type of __q_dict must be dict.")
        self.__q_dict = value

    q_dict = property(get_q_dict, set_q_dict)

    def extract_q_dict(self, state_key, action_key):
        '''
        Extract Q-Value from `self.q_dict`.

        Args:
            state_key:      The key of state.
            action_key:     The key of action.

        Returns:
            Q-Value.

        '''
        q = 0.0
        try:
            q = self.q_dict[state_key][action_key]
        except KeyError:
            self.save_q_dict(state_key, action_key, q)

        return q

    def save_q_dict(self, state_key, action_key, q_value):
        '''
        Insert or update Q-Value in `self.q_dict`.

        Args:
            state_key:      State.
            action_key:     Action.
            q_value:        Q-Value.

        Exceptions:
            TypeError:      If the type of `q_value` is not float.

        '''
        if isinstance(q_value, float) is False:
            raise TypeError("The type of q_value must be float.")

        if state_key not in self.q_dict:
            self.q_dict[state_key] = {action_key: q_value}
        else:
            self.q_dict[state_key][action_key] = q_value

    # R(state)
    __r_dict = {}

    def get_r_dict(self):
        '''
        getter
        R(state)
        '''
        if isinstance(self.__r_dict, dict) is False:
            raise TypeError("The type of __r_dict must be dict.")
        return self.__r_dict

    def set_r_dict(self, value):
        '''
        setter
        R(state)
        '''
        if isinstance(value, dict) is False:
            raise TypeError("The type of __r_dict must be dict.")
        self.__r_dict = value

    r_dict = property(get_r_dict, set_r_dict)

    def extract_r_dict(self, state_key, action_key=None):
        '''
        Extract R-Value from `self.r_dict`.

        Args:
            state_key:     The key of state.
            action_key:    The key of action.

        Returns:
            R-Value(Reward).

        '''
        try:
            if action_key is None:
                return self.r_dict[state_key]
            else:
                return self.r_dict[(state_key, action_key)]
        except KeyError:
            self.save_r_dict(state_key, 0.0, action_key)
            return self.extract_r_dict(state_key, action_key)

    def save_r_dict(self, state_key, r_value, action_key=None):
        '''
        Insert or update R-Value in `self.r_dict`.

        Args:
            state_key:     The key of state.
            r_value:       R-Value(Reward).
            action_key:    The key of action if it is nesesary for the parametar of value function.

        Exceptions:
            TypeError:      If the type of `r_value` is not float.
        '''
        if isinstance(r_value, float) is False:
            raise TypeError("The type of r_value must be float.")

        if action_key is not None:
            self.r_dict[(state_key, action_key)] = r_value
        else:
            self.r_dict[state_key] = r_value

    # Time.
    __t = 0

    def get_t(self):
        '''
        getter
        Time.
        '''
        if isinstance(self.__t, int) is False:
            raise TypeError("The type of __t must be int.")
        return self.__t

    def set_t(self, value):
        '''
        setter
        Time.
        '''
        if isinstance(value, int) is False:
            raise TypeError("The type of __t must be int.")
        self.__t = value

    t = property(get_t, set_t)

    @abstractmethod
    def learn(self, state_key, limit):
        '''
        Learning.
        
        Abstract method for concreate usecases.

        Args:
            state_key:      State.
            limit:          The number of learning.

        '''
        raise NotImplementedError("This method must be implemented.")

    @abstractmethod
    def select_action(self, state_key, next_action_list):
        '''
        Select action by Q(state, action).

        Abstract method for concreate usecases.

        Args:
            state_key:              The key of state.
            next_action_list:       The possible action in `self.t+1`.
                                    If the length of this list is zero, all action should be possible.

        Retruns:
            The key of action.

        '''
        raise NotImplementedError("This method must be implemented.")

    @abstractmethod
    def extract_possible_actions(self, state_key):
        '''
        Extract the list of the possible action in `self.t+1`.

        Abstract method for concreate usecases.

        Args:
            state_key       The key of state in `self.t+1`.

        Returns:
            The possible action in `self.t+1`.

        '''
        raise NotImplementedError("This method must be implemented.")

    def update_q(self, state_key, action_key, reward_value, next_max_q):
        '''
        Update Q-Value.

        Args:
            state_key:              The key of state.
            action_key:             The key of action.
            reward_value:           R-Value(Reward).
            next_max_q:             Maximum Q-Value.

        '''
        # Now Q-Value.
        q = self.extract_q_dict(state_key, action_key)
        # Update Q-Value.
        new_q = q + self.alpha_value * (reward_value + (self.gamma_value * next_max_q) - q)
        # Save updated Q-Value.
        self.save_q_dict(state_key, action_key, new_q)

    def predict_next_action(self, state_key, next_action_list):
        '''
        Predict next action by Q-Learning.

        Args:
            state_key:          The key of state in `self.t+1`.
            next_action_list:   The possible action in `self.t+1`.

        Returns:
            The key of action.

        '''
        next_action_q_list = [(action_key, self.extract_q_dict(state_key, action_key)) for action_key in next_action_list]
        random.shuffle(next_action_q_list)
        max_q_action = max(next_action_q_list, key=itemgetter(1))[0]

        return max_q_action

File: Reinforcement-Learning-with-js/jsqlearning/qlearning.js
/*
Repository name: accel-brain-code
Description: Base class and `Template Method Pattern` of Q-Learning.
Version: 1.0.1
Author: chimera0(RUM)
Author URI: http://accel-brain.com/
License: GNU General Public License v2.0
Copyright 2017 chimera0 (email : ai-brain-lab@accel-brain.com)
*/


var QLearning = (function()
{
    /**
     *
     * @private
     *
     */
    var strategy_ = null;

    /** 
     * @private
     *
     * Learning rate.
     */
    var alpha_value_ = 0.1;

    /** 
     * @private
     *
     * Gamma value.
     */
    var gamma_value_ = 0.5;

    /** 
     * @private
     *
     * Q(state, action).
     */
    var q_dict_ = {};

    /** 
     * @private
     *
     * R(state).
     */
    var r_dict_ = {};

    /** 
     * @private
     *
     * Time.
     */
    var t_ = 0;

    /**
     * Set Up hyperparams.
     *
     * @constructor
     * @params {object}
     */
    var constructor = function(strategy, params) {
        this.strategy_ = strategy;
        if (params != undefined)
        {
            if ("alpha_value" in params)
            {
                this.alpha_value = params.alpha_value;
            }
            if ("gamma_value" in params)
            {
                this.gamma_value = params.gamma_value;
            }
            if ("q_dict" in params)
            {
                this.q_dict = params.q_dict;
            }
        }
    }

    /** @constructor */
    constructor.prototype = {
        alpha_value: alpha_value_,
        gamma_value: gamma_value_,
        q_dict: q_dict_,
        r_dict: r_dict_,

        /**
         *
         * Extract Q-Value from `self.q_dict`.
         *
         * @params {string}
         * @params {string}
         *
         * @return {float}
         */
        extract_q_dict: function (state_key, action_key)
        {
            try
            {
                if (this.q_dict == undefined)
                {
                    this.save_q_dict(state_key, action_key, 0.0);
                    return 0.0;
                }
                else
                {
                    var state_key_list = Object.keys(this.q_dict);
                    if (state_key in this.q_dict && this.q_dict[state_key] != undefined)
                    {
                        if (action_key in this.q_dict[state_key])
                        {
                            return this.q_dict[state_key][action_key];
                        }
                        else
                        {
                            this.save_q_dict(state_key, action_key, 0.0);
                            return 0.0;
                        }
                    }
                    else
                    {
                        this.save_q_dict(state_key, action_key, 0.0);
                        return 0.0;
                    }
                }
            }
            catch (e)
            {
                console.log(e);
                return 0.0;
            }
        },

        /**
         *
         * Insert or update Q-Value in `self.q_dict`.
         *
         * @params {string}
         * @params {string}
         * @params {float}
         */
        save_q_dict: function(state_key, action_key, q_value)
        {
            if (q_value.isNaN == true || (q_value !== q_value) === true) return;
            if (state_key == false) return;
            if (action_key == false) return;
            if (q_value == false) return;
            if (this.q_dict == undefined) this.q_dict = {};
            if (this.q_dict[state_key] == undefined)
            {
                this.q_dict[state_key] = {};
            }
            this.q_dict[state_key][action_key] = q_value;
        },

        /**
         *
         * Extract R-Value from `self.r_dict`.
         *
         * @params {string}
         * @params {string}
         *
         * @return {float}
         */
        extract_r_dict : function (state_key, action_key)
        {
            try
            {
                if (this.r_dict == undefined)
                {
                    this.save_r_dict(state_key, 0.0, action_key);
                    return 0.0;
                }
                else
                {
                    if (state_key in this.r_dict && this.r_dict[state_key] != undefined)
                    {
                        if (action_key == undefined)
                        {
                            reward_value = this.r_dict[state_key];
                        }
                        else
                        {
                            if (action_key in this.r_dict[state_key])
                            {
                                reward_value = this.r_dict[state_key][action_key];
                            }
                            else
                            {
                                this.save_r_dict(state_key, 0.0, action_key);
                                return 0.0;
                            }
                        }
                    }
                    else
                    {
                        this.save_r_dict(state_key, 0.0, action_key);
                        return 0.0;
                    }
                }
            }
            catch(e)
            {
                console.log(e);
                return 0.0;
            }
        },

        /**
         *
         * Insert or update Reward Value in `self.r_dict`.
         *
         * @params {string}
         * @params {string}
         * @params {float}
         */
        save_r_dict: function (state_key, reward_value, action_key)
        {
            if (reward_value.isNaN == true || (reward_value !== reward_value) === true) return;
            if (state_key == false) return;
            if (reward_value == false) return;

            if (this.r_dict == undefined) this.r_dict = {};
            if (this.r_dict[state_key] == undefined)
            {
                if (action_key != undefined)
                {
                    this.r_dict[state_key] = {};
                }
                else
                {
                    this.r_dict[state_key] = 0.0;
                }
            }
            if (action_key != undefined)
            {
                this.r_dict[state_key][action_key] = reward_value;
            }
            else
            {
                this.r_dict[state_key] = reward_value;
            }
        },

        /**
         * Time.
         */
        t: t_,

        /**
         *
         * Learning.
         *
         * @params {string}
         * @params {int}
         */
        learn: function (state_key, limit)
        {
            if (limit == undefined) limit = 10000;
            this.t = 1;
            for (var _t = 1;_t<=limit;_t++)
            {
                next_action_list = this.extract_possible_actions(state_key);
                action_key = this.select_action(
                    state_key,
                    next_action_list
                )
                reward_value = this.observe_reward_value(state_key, action_key)
                // Vis.
                this.visualize_learning_result(state_key)
                // Check.
                if (this.check_the_end_flag(state_key) == true)
                {
                    break;
                }

                // Max-Q-Value in next action time.
                next_next_action_list = this.extract_possible_actions(action_key)
                next_action_key = this.predict_next_action(action_key, next_next_action_list)

                next_max_q = this.extract_q_dict(action_key, next_action_key)
                q = this.extract_q_dict(state_key, action_key);

                // Update Q-Value.
                this.update_q(
                    state_key,
                    action_key,
                    reward_value,
                    next_max_q
                );

                // Episode.
                this.t = _t;

                // Update State.
                state_key = this.update_state(
                    state_key,
                    action_key
                )
            }
        },

        /**
         * Select action by Q(state, action).
         *
         * @params{string}
         * @params{array}
         *
         * @return{string}
         */
        select_action: function (state_key, next_action_list)
        {
            return this.strategy_.select_action(this, state_key, next_action_list);
        },

        /**
         * Extract the list of the possible action in `self.t+1`.
         *
         * @params{string}
         *
         * @return{array}
         */
        extract_possible_actions: function (state_key)
        {
            return this.strategy_.extract_possible_actions(this, state_key);
        },

        /**
         * Extract the list of the possible action in `self.t+1`.
         *
         * @params{string}
         * @params{string}
         *
         * @return{array}
         */
        observe_reward_value: function (state_key, action_key)
        {
            return this.strategy_.observe_reward_value(this, state_key, action_key);
        },

        /**
         * Update Q-Value.
         *
         * @params{string}
         * @params{string}
         * @params{float}
         * @params{float}
         *
         */
        update_q: function (state_key, action_key, reward_value, next_max_q)
        {
            // Now Q-Value.
            var q = this.extract_q_dict(state_key, action_key);
            // Update Q-Value.
            var new_q = q + this.alpha_value * (reward_value + (this.gamma_value * next_max_q) - q);
            // Save updated Q-Value.
            this.save_q_dict(state_key, action_key, new_q);
        },

        /**
        * Predict next action by Q-Learning.
        *
        * @params{string}
        * @params{array}
        *
        * @return{string}
        */
        predict_next_action(state_key, next_action_list)
        {
            next_action_q_list = [];
            max_q = 0.0;
            max_q_action = null;
            for (var i=0; i<next_action_list.length;i++)
            {
                q = this.extract_q_dict(state_key, next_action_list[i]);
                if (max_q < q)
                {
                    max_q = q;
                    max_q_action = action_key;
                }
            }
            return max_q_action;
        },

        /**
         * Update state.
         *
         * @params{string}
         * @params{string}
         *
         * @return{string}
         */
        update_state: function(state_key, action_key)
        {
            if (this.strategy_.update_state == undefined)
            {
                return action_key;
            }
            else
            {
                return this.strategy_.update_state(this, state_key, action_key);
            }
        },
        
        /**
         * Check the end flag.
         *
         * If this return value is `True`, the learning is end.
         * As a rule, the learning can not be stopped.
         * This method should be overrided for concreate usecases.
         *
         * @params{string}
         *
         * @return{bool}
         */
        check_the_end_flag: function(state_key)
        {
            if (this.strategy_.check_the_end_flag == undefined)
            {
                return false;
            }
            else
            {
                return this.check_the_end_flag(this, state_key)
            }
        },
        
        /**
         *  Visualize learning result.
         *
         * @params{string}
         *
         */
        visualize_learning_result: function (state_key)
        {
            if (this.strategy_.visualize_learning_result != undefined)
            {
                this.strategy_.visualize_learning_result(this, state_key);
            }
        }
    };
    return constructor;
}
) ();
